### PR TITLE
adding postcode validation and token pre-processing. adding supportin…

### PR DESCRIPTION
**Pull request by iRhysBradbury**
…g tests

### CODE REVIEW

File: parsers/src/main/scala/uk/gov/ons/addressIndex/parsers/Tokens.scala [hunk 1]
```diff
@@ -55,11 +56,43 @@ object Tokens extends CrfTokenable {
...
   }
 
+  def validatePostCode(tokens: Map[String, Seq[CrfTokenResult]]): Map[String, Seq[CrfTokenResult]] = {
+    val postCodeTokens = tokens.getOrElse(Tokens.postcode, Seq.empty)
+    if(postCodeTokens.size == 1) {
+      val token = postCodeTokens.head.value
+      if(token.length >= 5) {
+        val outcode = token.substring(token.length - 3, token.length)
+        val incode = token.substring(0, token.indexOf(outcode))
+        val regex = "[0-9][A-Z][A-Z]".r
```
Comment (saniemi) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/parsers/Tokens.scala]: In the future we may wish to consider using
` [0-9][A-Z-[CIKMOV]]{2}`
or something more robust, but for time being this is fine.

File: parsers/src/main/scala/uk/gov/ons/addressIndex/parsers/Tokens.scala [hunk 2]
```diff
@@ -55,11 +56,43 @@ object Tokens extends CrfTokenable {
...
+    validatedTokens.map {
       case (token, values) => (token, values.map(_.value).mkString(" "))
     }
   }
 
+  def validatePostCode(tokens: Map[String, Seq[CrfTokenResult]]): Map[String, Seq[CrfTokenResult]] = {
+    val postCodeTokens = tokens.getOrElse(Tokens.postcode, Seq.empty)
+    if(postCodeTokens.size == 1) {
+      val token = postCodeTokens.head.value
+      if(token.length >= 5) {
```
Comment (saniemi) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/parsers/Tokens.scala]: Why 5? I thought that it should be >= 4?
"X1AA" -> "X 1AA"?

File: parsers/src/main/scala/uk/gov/ons/addressIndex/parsers/Tokens.scala [hunk 3]
```diff
@@ -55,11 +56,43 @@ object Tokens extends CrfTokenable {
...
+            key = Tokens.postcode,
+            value = Seq(
+              CrfTokenResult(
+                value = s"$incode $outcode",
+                label = Tokens.postcode
+              )
+            )
+          )
+        } else {
+          tokens - Tokens.postcode
```
Comment (saniemi) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/parsers/Tokens.scala]: My Scala is non existing, so apologies if this is irrelevant. I assume that the - operator here removes the postcode token from the sequence of tokens, correct? If so, then it's correct.

Comment (iRhysBradbury) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/parsers/Tokens.scala]: Correct :)

ref to func : http://www.scala-lang.org/api/2.12.x/scala/collection/immutable/Map.html#-(key:K):scala.collection.Map[K,V]

File: parsers/src/main/scala/uk/gov/ons/addressIndex/parsers/Tokens.scala (hunk 2 above)
Comment (iRhysBradbury) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/parsers/Tokens.scala]: Good spot!

File: parsers/src/main/scala/uk/gov/ons/addressIndex/parsers/Tokens.scala [hunk 4]
```diff
@@ -55,11 +56,43 @@ object Tokens extends CrfTokenable {
   override def normalise(tokens: Array[String]): Array[String] = tokens.map(_.toUpperCase)
 
   def tokensToMap(tokens: Seq[CrfTokenResult]): Map[String, String] = {
-    tokens.groupBy(_.label).map {
+    val validatedTokens = validatePostCode(tokens.groupBy(_.label))
```
Comment (Mironor) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/parsers/Tokens.scala]: The aim of the function `tokensToMap` is to transform a list of tokens into a `Map` with a label as a key and concatenated value as a value, why do we have a validation (of the postcode) inside of it?

Comment (iRhysBradbury) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/parsers/Tokens.scala]: Hi alex, we need to do this to all postcode tokens, all the time

Comment (Mironor) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/parsers/Tokens.scala]: Then it should be in the `apply` method, shouldn't it?

Comment (iRhysBradbury) [inline parsers/src/main/scala/uk/gov/ons/addressIndex/parsers/Tokens.scala]: It's a little different to that, this is transforming the tokens to valid map of tokens to be used in the query, changing func name

File: parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/PostCodeValidationTest.scala [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+  it should "X1AA -> X 1AA" in {
+    testHelper(
+      input = "X1AA",
+      expected = "X 1AA"
+    )
+  }
+
+  //These ingored tests are bounds tests for postcode, these occurrences will never happen
+  //because the parser will pick them up as building number.
+  //still interesting to keep.
```
Comment (Mironor) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/PostCodeValidationTest.scala]: If you want to keep ignored tests as a documentation, please add the information that they contain to the javadoc instead and remove the ignored tests

Comment (iRhysBradbury) [inline parsers/src/test/scala/uk/gov/ons/addressIndex/parsers/PostCodeValidationTest.scala]: This is a special request from @Ngwara

**Pull request merged**